django_short_urls/views.py

```diff
@@ -0,0 +1,176 @@
+_EXTRACT_VALID_PATH_RE = re.compile(r'^[%s]*' % URL_SAFE_FOR_PATH)
+
+
+def _extract_valid_path(path):
+    """Remove anything after the first non-URL_SAFE_FOR_PATH char as well as the last potential trailing '/',"""
+
+    path = _EXTRACT_VALID_PATH_RE.match(path).group(0)
+
+    if path[-1:] == '/':
+        # This can't be done directly in the regex because of greediness issues (URL_SAFE_FOR_PATH includes '/')
+        return path[:-1]
+
+    return path
+
+
+# pylint: disable=E1101, W0511
+@require_safe
+def main(request, path):
+    '''
+    Search for a long link matching the `path` and redirect
+    '''
+
+    #
+    path = _extract_valid_path(path)
+
+    link = Link.find_by_hash(path)
+
+    if link is None:
+        # Try to find a matching short link by removing valid "catchall" suffixes
+        path_prefix, redirect_suffix = suffix_catchall.get_hash_from(path)
+
+        if redirect_suffix is not None:
+            # If we found a suffix, we try to find a link again with the prefix
+            link = Link.find_by_hash(path_prefix)
+    else:
+        redirect_suffix = None
+
+    # Instrumentation
+    prefix_tag = 'prefix:' + link.prefix if link else 'Http404'
+
+    statsd.increment('workforus.clicks', tags=[prefix_tag])
+    statsd.set('workforus.unique_links', link.hash if link else 'Http404', tags=[prefix_tag])
+    statsd.set('workforus.unique_ips', get_client_ip(request), tags=['browser:' + get_browser(request)])
+
+    # 404 if link not found or register a click if the DB is not in readonly mode
+    if link is None:
+        raise Http404
+    elif mongoengine_is_primary():
+        link.click()
+
+    # Tweak the redirection link based on the query string, redirection suffix, etc.
+    # FIXME: Handle multiple parameters with the same name in the `url`
+    query = request.GET.copy()
+
+    if redirect_suffix is not None:
+        query[REDIRECT_PARAM_NAME] = redirect_suffix
+
+    if bool(query) and REF_PARAM_NAME not in query:
+        # If we specify a non empty query, indicate that the shortener tweaked the url
+        query[REF_PARAM_NAME] = REF_PARAM_DEFAULT_VALUE
+
+    target_url = url_append_parameters(
+        link.long_url,
+        params_to_replace=query,
+        defaults={REF_PARAM_NAME: REF_PARAM_DEFAULT_VALUE}
+    )
+
+    # FIXME: OPS-3885 - we're logging agressively to Sentry here because we don't understand some weird 404 errors on
+    # FIXME: workfor.us. We suspect they are being triggered by some pages being served directly as a proxy, perhaps.
+    if link.act_as_proxy:
+        logging.error('Page served through proxy: %s', str(link))
+
+    # Either redirect the user, or load the target page and display it directly
+    return (proxy if link.act_as_proxy else redirect)(target_url)
+
+
+@require_POST
+def new(request):
+    '''
+    Create a new short url based on the POST parameters
+    '''
+    user = None
+    if 'HTTP_AUTHORIZATION' in request.META:
```

Comment:
I'm… kind of mind-boggled that Django doesn't have any basic auth facilities.

That whole (pretty generic) basic auth section is certainly worth being in a separate helper.